[SME] Add agnostic-ZA interface and routines to save/restore SME state (#264)
ARM-software/abi-aa: aapcs64/aapcs64.rst

   * the current values of TPIDR2_EL0, PSTATE.SM and PSTATE.ZA.

>    * the registers enabled by PSTATE.ZA, if PSTATE.ZA is enabled.

Author: Sorry this should not have been committed. I've removed the change.

    requirements specified by ``OPTIONS``.

* The subroutine does not return a value.

* The subroutine behaves as follows:

  * The subroutine aborts in some platform-specific manner if either of the
    following conditions is true:

    * The current thread does not have access to SME.

    * The current thread does not have access to ``TPIDR2_EL0`` when
      PSTATE.ZA is 1.

  * If ``PTR`` does not point to a valid buffer with the required size, the
    behaviour of calling this routine is undefined.

  * If PSTATE.ZA is 0, the subroutine does nothing.

  * If bit 0 of ``OPTIONS`` is 0, then for addresses ``PTR->SAVED_ZA``,
    ``PTR->BLK``, ``PTR->ZA`` and ``PTR->TPIDR2_EL0`` at unspecified offsets
    in the buffer pointed to by ``PTR``:

    * The full contents of ``TPIDR2_EL0`` are written to ``PTR->TPIDR2_EL0``.

    * The address ``PTR->ZA`` is written to ``PTR->BLK.za_save_buffer``,
      the streaming vector length in bytes (``SVL.B``) is written to
      ``PTR->BLK.num_za_save_slices`` and the address ``PTR->BLK`` is
      written to ``TPIDR2_EL0``, thus setting up a lazy save.

    * The value 1 is written to ``PTR->SAVED_ZA``.

>   * If bit 1 of ``OPTIONS`` is 0 and ZT0 is available, then for the addresses

Author: Good spot, you're right!

  * Otherwise, the subroutine returns the value 0 in X0.

``__arm_sme_state_size``
^^^^^^^^^^^^^^^^^^^^^^^^

**(Beta)**

Platforms that support SME must provide a subroutine that returns a size
that is large enough to represent all state enabled by PSTATE.ZA.

* The subroutine is called ``__arm_sme_state_size``.

* The subroutine has an `agnostic-ZA`_ `streaming-compatible interface`_ with
  the following properties:

  * X1-X15, X19-X29 and SP are call-preserved.
  * Z0-Z31 are call-preserved.
  * P0-P15 are call-preserved.

* The subroutine takes no arguments.

* The subroutine returns an unsigned double word in X0 that represents
  a size in bytes that is large enough to represent all state enabled by
  PSTATE.ZA as well as any other state required for `__arm_sme_save`_ and
  `__arm_sme_restore`_.

  The exact layout used to calculate the size is unspecified.  The
  implementations of `__arm_sme_save`_ and `__arm_sme_restore`_ and
  `__arm_sme_state_size`_ must all assume the same layout.

  The size is guaranteed to be a multiple of 16.

* The subroutine behaves as follows:

>   * If the current thread has access to FEAT_SME and PSTATE.ZA is 1,

Reviewer: Maybe “ZA is active” (perhaps with the same kind of internal link as the comment below) now that the routines don't save anything when ZA is dormant.

    X0 contains the total size required to save and restore all SME state
    enabled by PSTATE.ZA.

  * Otherwise, X0 contains a size large enough to represent internal state
    required for `__arm_sme_save`_ and `__arm_sme_restore`_.

``__arm_sme_save``
^^^^^^^^^^^^^^^^^^

**(Beta)**

Platforms that support SME must provide a subroutine to save any state enabled
by PSTATE.ZA.

* The subroutine is called ``__arm_sme_save``.

* The subroutine has a custom ``ZA`` `streaming-compatible interface`_ with
  the following properties:

  * X1-X15, X19-X29 and SP are call-preserved.
  * Z0-Z31 are call-preserved.
  * P0-P15 are call-preserved.

* The subroutine takes the following arguments:

  PTR
    a 64-bit data pointer passed in X0 that points to a buffer which
    is guaranteed to have a size that is equal to or larger than the size
    returned by `__arm_sme_state_size`_. The pointer must be 16-byte aligned.

* The subroutine does not return a value.

* The subroutine behaves as follows:

>   * If ``PTR`` does not point to a valid buffer with the required size, the

Reviewer: Perhaps we should explicitly require the routine to “abort the program in some platform-defined manner” if ``PTR`` is not 16-byte aligned.  This would allow the low 4 bits to be used for future extensions.  Same for the restore routine.

This would then be “Otherwise, if …”

    behavior of calling this subroutine is undefined.

>   * If ZA state is 'active' on entry, then it is 'dormant' on normal return.
>     Otherwise the ZA state is unchanged.

Reviewer: ```suggestion
  * If ZA state is `active <ZA active state>`_ on entry, then it is
    `dormant <ZA dormant state>`_ on normal return.  Otherwise the ZA state
    is unchanged.
```

(untested) with a new internal target for “ZA dormant state”.

Author: For some reason, these links don't work for me when I generate the PDF (active or dormant), even though I've spelled them correctly.

Reviewer: OK, how about using:

```
   If the `ZA state is active`_ on entry, then it is dormant on return. …
```

with:

```
active
   .. _`ZA active state`:

   .. _`ZA state is active`:

```

Hopefully the active link is a good enough clue regarding “dormant”.

Same for ``__arm_sme_state_size``, in particular with “the ZA state”.

  * For the address ``PTR->VALID`` at an unspecified offset in the buffer,
    the value 0 is written to ``PTR->VALID`` and the subroutine returns, if

>     either of the following conditions is true:

Reviewer: ```suggestion
    one of the following conditions is true:
```

    ``PTR->BLK.num_za_save_slices`` and finally copying the address ``PTR->BLK``
    to ``TPIDR2_EL0``.

  * If ZT0 is available, then for the address ``PTR->ZT0`` at an unspecified
    offset in the buffer pointed to by ``PTR``, the contents of ZT0 are written
    to ``PTR->ZT0``.

  * The value 1 is written to ``PTR->VALID``.

``__arm_sme_restore``
^^^^^^^^^^^^^^^^^^^^^

**(Beta)**

Platforms that support SME must provide a subroutine to restore any state
enabled by PSTATE.ZA.

* The subroutine is called ``__arm_sme_restore``.

* The subroutine has a custom ``ZA`` `streaming-compatible interface`_ with
  the following properties:

  * X1-X15, X19-X29 and SP are call-preserved.
  * Z0-Z31 are call-preserved.
  * P0-P15 are call-preserved.

* The subroutine takes the following arguments:

  PTR
    a 64-bit data pointer passed in X0 that points to a buffer that
    is initialized by a call to `__arm_sme_save`_. The pointer must be 16-byte
    aligned.

* The subroutine does not return a value.

* The subroutine behaves as follows:

  * If ``PTR`` does not point to a valid buffer with the required size, the
    behavior of calling this routine is undefined.

>   * The ZA state on normal return is the same as the ZA state on entry to the

Reviewer: This is no longer true if ZA was dormant or off, since the save might have been committed between the calls.  Perhaps we should just delete this bullet point and deal with the various cases below.  E.g. the next bullet point deals with the case where ZA was dormant or off on entry to the save function.

    call to `__arm_sme_save`_ that was used to initialize the buffer
    pointed to by ``PTR``.

  * For the address ``PTR->VALID`` at an unspecified offset in the buffer,
    if the value stored at address ``PTR->VALID`` is 0, then the subroutine
    does nothing.

  * Otherwise, the subroutine aborts in some platform-specific manner if
    either of the following conditions is true:

    * The current thread does not have access to SME.

>     * ZA state is 'active' on entry.

Reviewer: ```suggestion
    * ZA state is `active <ZA active state>`_ on entry.
```
